File: go-vllm-api/internal/handler/agent/agent.go
package agent

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"go-vllm-api/internal/service"
	"go-vllm-api/internal/tools"

	"github.com/gin-gonic/gin"
	"go.uber.org/zap"
)

type AgentHandler struct {
	scheduler   *service.Scheduler
	gpuMonitor  *service.GPUMonitor
	toolExecutor *tools.ToolExecutor
	toolRegistry *tools.ToolRegistry
	logger      *zap.Logger
	httpClient  *http.Client
}

type AgentMessage struct {
	Role       string                   `json:"role"`
	Content    string                   `json:"content"`
	ToolCalls  []map[string]any         `json:"tool_calls,omitempty"`
	ToolCallID string                   `json:"tool_call_id,omitempty"`
	Name       string                   `json:"name,omitempty"`
}

type AgentRequest struct {
	Model         string         `json:"model"`
	Messages      []AgentMessage `json:"messages"`
	Tools         []string       `json:"tools,omitempty"`
	MaxIterations int            `json:"max_iterations"`
	Stream        bool           `json:"stream"`
	AutoConfirm   bool           `json:"auto_confirm"`
}

type ToolCallRequest struct {
	Name      string         `json:"name"`
	Arguments map[string]any `json:"arguments"`
}

type ExecuteToolsRequest struct {
	Calls       []ToolCallRequest `json:"calls"`
	AutoConfirm bool              `json:"auto_confirm"`
}

func NewAgentHandler(scheduler *service.Scheduler, gpuMonitor *service.GPUMonitor, toolExecutor *tools.ToolExecutor, toolRegistry *tools.ToolRegistry, logger *zap.Logger) *AgentHandler {
	return &AgentHandler{
		scheduler:    scheduler,
		gpuMonitor:   gpuMonitor,
		toolExecutor: toolExecutor,
		toolRegistry: toolRegistry,
		logger:       logger,
		httpClient: &http.Client{
			Timeout: 120 * time.Second,
		},
	}
}

func (h *AgentHandler) RegisterRoutes(rg *gin.RouterGroup) {
	agent := rg.Group("/manage/agent")
	{
		agent.GET("/tools", h.ListTools)
		agent.GET("/tools/:tool_name", h.GetToolInfo)
		agent.POST("/execute", h.ExecuteTools)
		agent.POST("/execute/:tool_name", h.ExecuteSingleTool)
		agent.GET("/history", h.GetExecutionHistory)
		agent.DELETE("/history", h.ClearExecutionHistory)
		agent.POST("/chat", h.AgentChat)
	}
}

func (h *AgentHandler) ListTools(c *gin.Context) {
	categories := c.Query("categories")

	var toolsList []map[string]any
	if categories != "" {
		categoryList := strings.Split(categories, ",")
		for i, cat := range categoryList {
			categoryList[i] = strings.TrimSpace(cat)
		}
		toolsList = h.toolRegistry.GetOpenAITools(categoryList)
	} else {
		toolsList = h.toolRegistry.GetOpenAITools(nil)
	}

	c.JSON(http.StatusOK, gin.H{
		"tools":      toolsList,
		"categories": h.toolRegistry.GetCategories(),
		"count":      len(toolsList),
	})
}

func (h *AgentHandler) GetToolInfo(c *gin.Context) {
	toolName := c.Param("tool_name")
	tool := h.toolRegistry.Get(toolName)

	if tool == nil {
		c.JSON(http.StatusNotFound, gin.H{"error": fmt.Sprintf("Tool '%s' not found", toolName)})
		return
	}

	parameters := make([]map[string]any, 0, len(tool.Parameters))
	for _, p := range tool.Parameters {
		paramMap := map[string]any{
			"name":        p.Name,
			"type":        string(p.Type),
			"description": p.Description,
			"required":    p.Required,
		}
		if p.Default != nil {
			paramMap["default"] = p.Default
		}
		if len(p.Enum) > 0 {
			paramMap["enum"] = p.Enum
		}
		parameters = append(parameters, paramMap)
	}

	c.JSON(http.StatusOK, gin.H{
		"name":                 tool.Name,
		"description":          tool.Description,
		"parameters":           parameters,
		"category":             tool.Category,
		"dangerous":            tool.Dangerous,
		"requires_confirmation": tool.RequiresConfirmation,
		"openai_schema":        tool.ToOpenAIFunction(),
	})
}

func (h *AgentHandler) ExecuteTools(c *gin.Context) {
	var req ExecuteToolsRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	calls := make([]tools.ToolCall, 0, len(req.Calls))
	for _, call := range req.Calls {
		calls = append(calls, tools.ToolCall{
			Name:      call.Name,
			Arguments: call.Arguments,
		})
	}

	results := h.toolExecutor.ExecuteBatch(c.Request.Context(), calls, req.AutoConfirm)

	resultList := make([]map[string]any, 0, len(results))
	allSuccess := true
	for _, r := range results {
		resultList = append(resultList, map[string]any{
			"tool_name":      r.ToolName,
			"success":        r.Success,
			"result":         r.Result,
			"error":          r.Error,
			"execution_time": r.ExecutionTime,
			"timestamp":      r.Timestamp,
		})
		if !r.Success {
			allSuccess = false
		}
	}

	c.JSON(http.StatusOK, gin.H{
		"results": resultList,
		"success": allSuccess,
	})
}

func (h *AgentHandler) ExecuteSingleTool(c *gin.Context) {
	_ = c.Param("tool_name")
	var req ToolCallRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	autoConfirm := false
	result := h.toolExecutor.Execute(c.Request.Context(), req.Name, req.Arguments, autoConfirm)

	c.JSON(http.StatusOK, gin.H{
		"tool_name":      result.ToolName,
		"success":        result.Success,
		"result":         result.Result,
		"error":          result.Error,
		"execution_time": result.ExecutionTime,
		"timestamp":      result.Timestamp,
	})
}

func (h *AgentHandler) GetExecutionHistory(c *gin.Context) {
	limit := 100
	if l := c.Query("limit"); l != "" {
		fmt.Sscanf(l, "%d", &limit)
	}

	history := h.toolExecutor.GetHistory(limit)
	stats := h.toolExecutor.GetStatistics()

	historyList := make([]map[string]any, 0, len(history))
	for _, r := range history {
		historyList = append(historyList, map[string]any{
			"tool_name":      r.ToolName,
			"success":        r.Success,
			"result":         r.Result,
			"error":          r.Error,
			"execution_time": r.ExecutionTime,
			"timestamp":      r.Timestamp,
		})
	}

	c.JSON(http.StatusOK, gin.H{
		"history":    historyList,
		"statistics": stats,
	})
}

func (h *AgentHandler) ClearExecutionHistory(c *gin.Context) {
	h.toolExecutor.ClearHistory()
	c.JSON(http.StatusOK, gin.H{"status": "success", "message": "History cleared"})
}

func (h *AgentHandler) AgentChat(c *gin.Context) {
	var req AgentRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	modelName := req.Model
	if modelName == "" {
		modelName = h.scheduler.GetDefaultModel()
	}
	if modelName == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "No model specified and no default set"})
		return
	}

	if !h.scheduler.IsModelAvailable(modelName) {
		c.JSON(http.StatusNotFound, gin.H{"error": fmt.Sprintf("Model not found: %s", modelName)})
		return
	}

	if !h.scheduler.IsModelRunning(modelName) {
		ok, err := h.scheduler.StartModel(c.Request.Context(), modelName)
		if !ok {
			c.JSON(http.StatusServiceUnavailable, gin.H{"error": fmt.Sprintf("Failed to start model: %v", err)})
			return
		}
		time.Sleep(5 * time.Second)
	}

	var toolsList []string
	if len(req.Tools) > 0 {
		toolsList = req.Tools
	}
	openAITools := h.toolRegistry.GetOpenAITools(toolsList)

	vllmPort := h.scheduler.GetModelPort(modelName)
	backendURL := fmt.Sprintf("http://localhost:%d", vllmPort)
	vllmModelName := h.scheduler.GetModelPath(modelName)
	if vllmModelName == "" {
		vllmModelName = modelName
	}

	h.logger.Info("agent chat request",
		zap.String("model", modelName),
		zap.String("vllm_model", vllmModelName),
		zap.String("backend_url", backendURL))

	if req.Stream {
		c.Header("Content-Type", "text/event-stream")
		c.Header("Cache-Control", "no-cache")
		c.Header("Connection", "keep-alive")
		c.Stream(func(w io.Writer) bool {
			ch := make(chan string, 1)
			go func() {
				h.agentStream(c.Request.Context(), backendURL, vllmModelName, req.Messages, openAITools, req.MaxIterations, req.AutoConfirm, ch)
				close(ch)
			}()

			for data := range ch {
				fmt.Fprint(w, data)
				if f, ok := w.(http.Flusher); ok {
					f.Flush()
				}
			}
			return false
		})
		return
	}

	result := h.agentCompletion(c.Request.Context(), backendURL, vllmModelName, req.Messages, openAITools, req.MaxIterations, req.AutoConfirm)
	c.JSON(http.StatusOK, result)
}

func (h *AgentHandler) agentCompletion(ctx context.Context, backendURL, modelName string, messages []AgentMessage, toolsList []map[string]any, maxIterations int, autoConfirm bool) map[string]any {
	openaiMessages := h.convertMessagesToOpenAI(messages)
	iteration := 0

	for iteration < maxIterations {
		iteration++

		payload := map[string]any{
			"model":       modelName,
			"messages":    openaiMessages,
			"tools":       toolsList,
			"tool_choice": "auto",
			"max_tokens":  2048,
		}

		data, err := h.postToVLLM(ctx, backendURL, "/v1/chat/completions", payload)
		if err != nil {
			return map[string]any{"error": fmt.Sprintf("Model API error: %s", err.Error())}
		}

		choices, ok := data["choices"].([]any)
		if !ok || len(choices) == 0 {
			return map[string]any{"error": "No choices in response"}
		}

		message, ok := choices[0].(map[string]any)["message"].(map[string]any)
		if !ok {
			return map[string]any{"error": "No message in response"}
		}

		toolCalls, hasToolCalls := message["tool_calls"].([]any)
		if !hasToolCalls || len(toolCalls) == 0 {
			return map[string]any{
				"id":        data["id"],
				"model":     modelName,
				"message":   message,
				"iterations": iteration,
				"finished":  true,
			}
		}

		openaiMessages = append(openaiMessages, message)

		calls := make([]tools.ToolCall, 0, len(toolCalls))
		for _, tc := range toolCalls {
			toolCall, ok := tc.(map[string]any)
			if !ok {
				continue
			}
			funcInfo, ok := toolCall["function"].(map[string]any)
			if !ok {
				continue
			}
			name, _ := funcInfo["name"].(string)
			argsStr, _ := funcInfo["arguments"].(string)

			var args map[string]any
			if argsStr != "" {
				if err := json.Unmarshal([]byte(argsStr), &args); err != nil {
					args = make(map[string]any)
				}
			} else {
				args = make(map[string]any)
			}

			calls = append(calls, tools.ToolCall{
				Name:      name,
				Arguments: args,
			})

			results := h.toolExecutor.ExecuteBatch(ctx, calls, autoConfirm)

			for i, result := range results {
				toolCallID := ""
				if i < len(toolCalls) {
					if tc, ok := toolCalls[i].(map[string]any); ok {
						toolCallID, _ = tc["id"].(string)
					}
				}

				content := ""
				if result.Success {
					if jsonData, err := json.Marshal(result.Result); err == nil {
						content = string(jsonData)
					} else {
						content = fmt.Sprintf("%v", result.Result)
					}
				} else {
					if errData, err := json.Marshal(map[string]any{"error": result.Error}); err == nil {
						content = string(errData)
					} else {
						content = result.Error
					}
				}

				openaiMessages = append(openaiMessages, map[string]any{
					"role":         "tool",
					"tool_call_id": toolCallID,
					"content":      content,
				})
			}
		}
	}

	return map[string]any{"error": "Max iterations reached", "iterations": iteration}
}

func (h *AgentHandler) agentStream(ctx context.Context, backendURL, modelName string, messages []AgentMessage, toolsList []map[string]any, maxIterations int, autoConfirm bool, ch chan<- string) {
	openaiMessages := h.convertMessagesToOpenAI(messages)
	iteration := 0

	for iteration < maxIterations {
		iteration++

		payload := map[string]any{
			"model":       modelName,
			"messages":    openaiMessages,
			"tools":       toolsList,
			"tool_choice": "auto",
			"max_tokens":  2048,
		}

		data, err := h.postToVLLM(ctx, backendURL, "/v1/chat/completions", payload)
		if err != nil {
			ch <- fmt.Sprintf("data: %s\n\n", h.mustJSON(map[string]any{"error": err.Error()}))
			return
		}

		choices, ok := data["choices"].([]any)
		if !ok || len(choices) == 0 {
			ch <- fmt.Sprintf("data: %s\n\n", h.mustJSON(map[string]any{"error": "No choices in response"}))
			return
		}

		message, ok := choices[0].(map[string]any)["message"].(map[string]any)
		if !ok {
			ch <- fmt.Sprintf("data: %s\n\n", h.mustJSON(map[string]any{"error": "No message in response"}))
			return
		}

		toolCalls, hasToolCalls := message["tool_calls"].([]any)
		if !hasToolCalls || len(toolCalls) == 0 {
			payload["stream"] = true
			h.streamFromVLLM(ctx, backendURL, "/v1/chat/completions", payload, ch)
			return
		}

		ch <- fmt.Sprintf("data: %s\n\n", h.mustJSON(map[string]any{
			"type":  "tool_calls_start",
			"calls": len(toolCalls),
		}))

		openaiMessages = append(openaiMessages, message)

		for _, tc := range toolCalls {
			toolCall, ok := tc.(map[string]any)
			if !ok {
				continue
			}
			funcInfo, ok := toolCall["function"].(map[string]any)
			if !ok {
				continue
			}
			name, _ := funcInfo["name"].(string)
			argsStr, _ := funcInfo["arguments"].(string)
			tcID, _ := toolCall["id"].(string)

			ch <- fmt.Sprintf("data: %s\n\n", h.mustJSON(map[string]any{
				"type": "tool_executing",
				"name": name,
				"id":   tcID,
			}))

			var args map[string]any
			if argsStr != "" {
				if err := json.Unmarshal([]byte(argsStr), &args); err != nil {
					args = make(map[string]any)
				}
			} else {
				args = make(map[string]any)
			}

			result := h.toolExecutor.Execute(ctx, name, args, autoConfirm)

			ch <- fmt.Sprintf("data: %s\n\n", h.mustJSON(map[string]any{
				"type":    "tool_result",
				"name":    name,
				"id":      tcID,
				"success": result.Success,
				"result":  result.Result,
				"error":   result.Error,
			}))

			content := ""
			if result.Success {
				if jsonData, err := json.Marshal(result.Result); err == nil {
					content = string(jsonData)
				} else {
					content = fmt.Sprintf("%v", result.Result)
				}
			} else {
				if errData, err := json.Marshal(map[string]any{"error": result.Error}); err == nil {
					content = string(errData)
				} else {
					content = result.Error
				}
			}

			openaiMessages = append(openaiMessages, map[string]any{
				"role":         "tool",
				"tool_call_id": tcID,
				"content":      content,
			})
		}

		ch <- fmt.Sprintf("data: %s\n\n", h.mustJSON(map[string]any{
			"type":      "tool_calls_end",
			"calls":     len(toolCalls),
			"iteration": iteration,
		}))
	}

	ch <- fmt.Sprintf("data: %s\n\n", h.mustJSON(map[string]any{"error": "Max iterations reached"}))
}

func (h *AgentHandler) postToVLLM(ctx context.Context, baseURL, path string, payload map[string]any) (map[string]any, error) {
	url := baseURL + path
	payloadBytes, err := json.Marshal(payload)
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewBuffer(payloadBytes))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := h.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("VLLM returned status %d", resp.StatusCode)
	}

	var result map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}

	return result, nil
}

func (h *AgentHandler) streamFromVLLM(ctx context.Context, baseURL, path string, payload map[string]any, ch chan<- string) {
	url := baseURL + path
	payloadBytes, err := json.Marshal(payload)
	if err != nil {
		ch <- fmt.Sprintf("data: %s\n\n", h.mustJSON(map[string]any{"error": err.Error()}))
		return
	}

	req, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewBuffer(payloadBytes))
	if err != nil {
		ch <- fmt.Sprintf("data: %s\n\n", h.mustJSON(map[string]any{"error": err.Error()}))
		return
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "text/event-stream")

	resp, err := h.httpClient.Do(req)
	if err != nil {
		ch <- fmt.Sprintf("data: %s\n\n", h.mustJSON(map[string]any{"error": err.Error()}))
		return
	}
	defer resp.Body.Close()

	scanner := bufio.NewScanner(resp.Body)
	for scanner.Scan() {
		line := scanner.Text()
		if strings.HasPrefix(line, "data: ") {
			ch <- line + "\n\n"
		}
	}
}

func (h *AgentHandler) convertMessagesToOpenAI(messages []AgentMessage) []map[string]any {
	result := make([]map[string]any, 0, len(messages))
	for _, msg := range messages {
		item := map[string]any{
			"role":    msg.Role,
			"content": msg.Content,
		}
		if len(msg.ToolCalls) > 0 {
			item["tool_calls"] = msg.ToolCalls
		}
		if msg.ToolCallID != "" {
			item["tool_call_id"] = msg.ToolCallID
		}
		if msg.Name != "" {
			item["name"] = msg.Name
		}
		result = append(result, item)
	}
	return result
}

func (h *AgentHandler) mustJSON(v any) string {
	data, err := json.Marshal(v)
	if err != nil {
		return fmt.Sprintf(`{"error":"json marshal failed: %s"}`, err.Error())
	}
	return string(data)
}

File: app-controller/routes/agent.py
"""Agent API Routes

Provides endpoints for AI agent interactions with tool calling support.
"""
from fastapi import APIRouter, HTTPException, Request
from fastapi.responses import StreamingResponse
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field
from datetime import datetime
import asyncio
import json
import os

from tools.definitions import ToolRegistry
from tools.executor import ToolExecutor, ToolResult
from middleware.error_handler import ModelNotFoundException, ModelServiceUnavailableException


agent_router = APIRouter(prefix="/manage/agent")


def _session_mgr():
    import main as m
    return m.agent_session_manager


# Pydantic models for request/response
class AgentMessage(BaseModel):
    role: str = Field(..., description="Message role: user, assistant, or system")
    content: str = Field(..., description="Message content")
    tool_calls: Optional[List[Dict[str, Any]]] = None
    tool_call_id: Optional[str] = None
    name: Optional[str] = None


class AgentRequest(BaseModel):
    model: Optional[str] = None
    messages: List[AgentMessage]
    tools: Optional[List[str]] = Field(None, description="Tool categories to enable")
    max_iterations: int = Field(5, description="Maximum tool call iterations")
    stream: bool = Field(True, description="Enable streaming response")
    auto_confirm: bool = Field(False, description="Auto-confirm dangerous tools")


class ToolCallRequest(BaseModel):
    name: str
    arguments: Dict[str, Any]


class ExecuteToolsRequest(BaseModel):
    calls: List[ToolCallRequest]
    auto_confirm: bool = False


def _scheduler():
    import main as m
    return m.scheduler


def _gpu_monitor():
    import main as m
    return m.gpu_monitor


def _metrics():
    import main as m
    return m.metrics


def _get_executor() -> ToolExecutor:
    """Get or create tool executor instance"""
    return ToolExecutor()


def _get_registry() -> ToolRegistry:
    """Get tool registry singleton"""
    return ToolRegistry()


@agent_router.get("/tools")
async def list_tools(categories: Optional[str] = None):
    """List all available tools or tools in specific categories"""
    registry = _get_registry()

    if categories:
        category_list = [c.strip() for c in categories.split(",")]
        tools = []
        for cat in category_list:
            tools.extend(registry.get_by_category(cat))
    else:
        tools = registry.get_all()

    return {
        "tools": [t.to_openai_function() for t in tools],
        "categories": registry.get_categories(),
        "count": len(tools)
    }


@agent_router.get("/tools/{tool_name}")
async def get_tool_info(tool_name: str):
    """Get detailed information about a specific tool"""
    registry = _get_registry()
    tool = registry.get(tool_name)

    if not tool:
        raise HTTPException(status_code=404, detail=f"Tool '{tool_name}' not found")

    return {
        "name": tool.name,
        "description": tool.description,
        "parameters": [p.__dict__ for p in tool.parameters],
        "category": tool.category,
        "dangerous": tool.dangerous,
        "requires_confirmation": tool.requires_confirmation,
        "openai_schema": tool.to_openai_function()
    }


@agent_router.post("/execute")
async def execute_tool(request: ExecuteToolsRequest):
    """Execute one or more tool calls"""
    executor = _get_executor()

    calls = [{"name": c.name, "arguments": c.arguments} for c in request.calls]
    results = await executor.execute_batch(calls, request.auto_confirm)

    return {
        "results": [r.to_dict() for r in results],
        "success": all(r.success for r in results)
    }


@agent_router.post("/execute/{tool_name}")
async def execute_single_tool(tool_name: str, request: ToolCallRequest):
    """Execute a single tool call"""
    executor = _get_executor()

    result = await executor.execute(tool_name, request.arguments, request.auto_confirm)
    return result.to_dict()


@agent_router.get("/history")
async def get_execution_history(limit: int = 100):
    """Get tool execution history"""
    executor = _get_executor()
    return {
        "history": executor.get_history(limit),
        "statistics": executor.get_statistics()
    }


@agent_router.delete("/history")
async def clear_execution_history():
    """Clear tool execution history"""
    executor = _get_executor()
    executor.clear_history()
    return {"status": "success", "message": "History cleared"}


@agent_router.post("/chat")
async def agent_chat(request: Request, body: AgentRequest):
    """Agent chat endpoint with automatic tool execution

    This implements an agentic loop:
    1. Send user message to model with available tools
    2. If model returns tool calls, execute them
    3. Feed tool results back to model
    4. Repeat until model returns final response or max iterations
    """
    scheduler = _scheduler()
    registry = _get_registry()
    executor = _get_executor()

    model_name = body.model or scheduler.get_default_model()
    if not model_name:
        raise HTTPException(status_code=400, detail="No model specified and no default set")

    if not scheduler.is_model_available(model_name):
        raise ModelNotFoundException(model_name)

    # Ensure model is running
    if not scheduler.is_model_running(model_name):
        success = await scheduler.start_model(model_name)
        if not success:
            raise ModelServiceUnavailableException(model_name, "Failed to start model")
        await asyncio.sleep(5)

    # Get tools for the request
    if body.tools:
        tools = registry.get_openai_tools(body.tools)
    else:
        tools = registry.get_openai_tools()

    vllm_port = scheduler.get_model_port(model_name)
    backend_url = f"http://localhost:{vllm_port}"
    vllm_model_name = scheduler.get_model_path(model_name) or model_name
    print(f"[DEBUG] model_name={model_name}, vllm_model_name={vllm_model_name}, backend_url={backend_url}")

    if body.stream:
        return StreamingResponse(
            _agent_stream(
                backend_url, vllm_model_name, body.messages, tools, executor,
                body.max_iterations, body.auto_confirm
            ),
            media_type="text/event-stream",
            headers={
                "Cache-Control": "no-cache",
                "Connection": "keep-alive"
            }
        )
    else:
        result = await _agent_completion(
            backend_url, vllm_model_name, body.messages, tools, executor,
            body.max_iterations, body.auto_confirm
        )
        return result


async def _agent_completion(
    backend_url: str,
    model_name: str,
    messages: List[AgentMessage],
    tools: List[Dict],
    executor: ToolExecutor,
    max_iterations: int,
    auto_confirm: bool
) -> Dict[str, Any]:
    """Non-streaming agent completion with tool calling loop"""
    import httpx

    client = httpx.AsyncClient(timeout=httpx.Timeout(120.0, connect=10.0))
    openai_messages = _convert_messages(messages)
    iteration = 0

    while iteration < max_iterations:
        iteration += 1

        payload = {
            "model": model_name,
            "messages": openai_messages,
            "tools": tools,
            "tool_choice": "auto",
            "max_tokens": 2048
        }

        try:
            response = await client.post(f"{backend_url}/v1/chat/completions", json=payload)
            response.raise_for_status()
            data = response.json()
        except httpx.HTTPError as e:
            return {"error": f"Model API error: {str(e)}"}

        message = data.get("choices", [{}])[0].get("message", {})
        tool_calls = message.get("tool_calls")

        if not tool_calls:
            # No more tool calls, return final response
            return {
                "id": data.get("id"),
                "model": model_name,
                "message": message,
                "iterations": iteration,
                "finished": True
            }

        # Add assistant message with tool calls to history
        openai_messages.append(message)

        # Execute tool calls
        calls = []
        for tc in tool_calls:
            args = tc.get("function", {}).get("arguments", "{}")
            if isinstance(args, str):
                try:
                    args = json.loads(args)
                except json.JSONDecodeError:
                    args = {}
            calls.append({
                "name": tc.get("function", {}).get("name"),
                "arguments": args,
                "id": tc.get("id")
            })

        results = await executor.execute_batch(
            [{"name": c["name"], "arguments": c["arguments"]} for c in calls],
            auto_confirm
        )

        # Add tool results to messages
        for call, result in zip(calls, results):
            openai_messages.append({
                "role": "tool",
                "tool_call_id": call["id"],
                "content": json.dumps(result.result) if result.success else json.dumps({"error": result.error})
            })

    return {"error": "Max iterations reached", "iterations": iteration}


async def _agent_stream(
    backend_url: str,
    model_name: str,
    messages: List[AgentMessage],
    tools: List[Dict],
    executor: ToolExecutor,
    max_iterations: int,
    auto_confirm: bool
):
    """Streaming agent completion with tool calling loop"""
    import httpx

    client = httpx.AsyncClient(timeout=httpx.Timeout(300.0, connect=10.0))
    openai_messages = _convert_messages(messages)
    iteration = 0

    while iteration < max_iterations:
        iteration += 1

        # First get non-streaming response to check for tool calls
        payload = {
            "model": model_name,
            "messages": openai_messages,
            "tools": tools,
            "tool_choice": "auto",
            "max_tokens": 2048
        }

        try:
            response = await client.post(f"{backend_url}/v1/chat/completions", json=payload)
            response.raise_for_status()
            data = response.json()
        except httpx.HTTPError as e:
            yield f"data: {json.dumps({'error': str(e)})}\n\n"
            return

        message = data.get("choices", [{}])[0].get("message", {})
        tool_calls = message.get("tool_calls")

        if not tool_calls:
            # Stream the final response
            payload["stream"] = True
            async with client.stream("POST", f"{backend_url}/v1/chat/completions", json=payload) as resp:
                async for line in resp.aiter_lines():
                    if line.startswith("data: "):
                        yield f"{line}\n\n"
            return

        # Notify about tool calls
        yield f"data: {json.dumps({'type': 'tool_calls_start', 'calls': len(tool_calls)})}\n\n"

        # Add assistant message to history
        openai_messages.append(message)

        # Execute tools and stream results
        for tc in tool_calls:
            func = tc.get("function", {})
            name = func.get("name", "unknown")
            args_str = func.get("arguments", "{}")
            tc_id = tc.get("id")

            try:
                args = json.loads(args_str) if isinstance(args_str, str) else args_str
            except json.JSONDecodeError:
                args = {}

            # Notify tool execution
            yield f"data: {json.dumps({'type': 'tool_executing', 'name': name, 'id': tc_id})}\n\n"

            result = await executor.execute(name, args, auto_confirm)

            # Send tool result
            yield f"data: {json.dumps({'type': 'tool_result', 'name': name, 'id': tc_id, 'success': result.success, 'result': result.result if result.success else None, 'error': result.error})}\n\n"

            # Add to message history
            openai_messages.append({
                "role": "tool",
                "tool_call_id": tc_id,
                "content": json.dumps(result.result) if result.success else json.dumps({"error": result.error})
            })

        yield f"data: {json.dumps({'type': 'tool_calls_end', 'calls': len(tool_calls), 'iteration': iteration})}\n\n"

    yield f"data: {json.dumps({'error': 'Max iterations reached'})}\n\n"


def _convert_messages(messages: List[AgentMessage]) -> List[Dict]:
    """Convert AgentMessage list to OpenAI format"""
    result = []
    for msg in messages:
        item = {"role": msg.role, "content": msg.content}
        if msg.tool_calls:
            item["tool_calls"] = msg.tool_calls
        if msg.tool_call_id:
            item["tool_call_id"] = msg.tool_call_id
        if msg.name:
            item["name"] = msg.name
        result.append(item)
    return result


# ===== Session Management =====

class CreateSessionRequest(BaseModel):
    title: str = Field("", description="Session title")
    model: Optional[str] = Field(None, description="Model to use")


class AddMessageRequest(BaseModel):
    role: str = Field(..., description="Message role")
    content: str = Field(..., description="Message content")


@agent_router.post("/sessions")
async def create_session(request: CreateSessionRequest):
    mgr = _session_mgr()
    session = mgr.create_session(title=request.title, model=request.model)
    return session.to_dict()


@agent_router.get("/sessions")
async def list_sessions(archived: Optional[bool] = None):
    mgr = _session_mgr()
    return {"sessions": mgr.list_sessions(archived=archived), "total": len(mgr.list_sessions(archived=archived))}


@agent_router.get("/sessions/{session_id}")
async def get_session(session_id: str):
    mgr = _session_mgr()
    session = mgr.get_session(session_id)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    return session.to_dict_full()


@agent_router.post("/sessions/{session_id}/messages")
async def add_session_message(session_id: str, request: AddMessageRequest):
    mgr = _session_mgr()
    session = mgr.add_message(session_id, request.role, request.content)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    return session.to_dict()


@agent_router.get("/sessions/{session_id}/messages")
async def get_session_messages(session_id: str, limit: int = 100):
    mgr = _session_mgr()
    messages = mgr.get_session_messages(session_id, limit)
    if messages is None:
        raise HTTPException(status_code=404, detail="Session not found")
    return {"messages": messages, "session_id": session_id}


@agent_router.delete("/sessions/{session_id}/messages")
async def clear_session_messages(session_id: str):
    mgr = _session_mgr()
    if not mgr.clear_session_messages(session_id):
        raise HTTPException(status_code=404, detail="Session not found")
    return {"status": "success", "message": "Messages cleared"}


@agent_router.delete("/sessions/{session_id}")
async def delete_session(session_id: str):
    mgr = _session_mgr()
    if not mgr.delete_session(session_id):
        raise HTTPException(status_code=404, detail="Session not found")
    return {"status": "success", "session_id": session_id}


@agent_router.post("/sessions/{session_id}/archive")
async def archive_session(session_id: str):
    mgr = _session_mgr()
    session = mgr.archive_session(session_id)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    return session.to_dict()
